DAOS-18019 test: Update to python3.11

phender opened the pull request:
Build and test with python3.11.

Skip-func-test-leap15: false
Skip-func-test-el9: false

### Steps for the author:

* [ ] Commit message follows the [guidelines](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Appropriate [Features or Test-tag](https://daosio.atlassian.net/wiki/spaces/DC/pages/10984259629/Test+Tags) pragmas were used.
* [ ] Appropriate [Functional Test Stages](https://daosio.atlassian.net/wiki/spaces/DC/pages/12147556353/CI+Functional+Test+Stages) were run.
* [ ] At least two positive code reviews including at least one code owner from each category referenced in the PR.
* [ ] Testing is complete. If necessary, forced-landing label added and a reason added in a comment.

#### After all prior steps are complete:
* [ ] Gatekeeper requested (daos-gatekeeper added as a reviewer).

phender committed:
DAOS-18019 test: Update to python3.11

Build and test with python3.11.

Skip-func-test-leap15: false
Skip-func-test-el9: false

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

phender committed:
Fix linting issues.

Skip-func-test-leap15: false
Skip-func-test-el9: false

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

phender committed:
Fix boost-python3-devel

Skip-func-test-leap15: false
Skip-func-test-el9: false

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

phender committed:
Fix linting issues.

Skip-func-test-leap15: false
Skip-func-test-el9: false

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

daosbuild3 commented:
Test stage Functional on EL 9 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/4/execution/node/1049/log

phender commented:
https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos/job/PR-16919/4/ completed all unit testing and ran all three functional VM stages.  The Functional on EL 9 stage did encounter the https://daosio.atlassian.net/browse/DAOS-17758 known issue with the `daos_test/dfuse.py` test.  All other pr, daily, and weekly tagged VM functional tests passed.

phender committed:
Merge branch 'master' into hendersp/DAOS-18019

phender committed:
Updates

Skip-func-test-leap15: false
Skip-func-test-el9: false
Test-tag: pr daily_regression

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

daosbuild3 commented:
Test stage Functional on Leap 15.6 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos/job/PR-16919/5/display/redirect

phender committed:
Additional updates.

Skip-func-test-leap15: false
Skip-func-test-el9: false
Test-tag: pr daily_regression

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

daosbuild3 commented:
Test stage Functional on EL 9 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/8/execution/node/1048/log

daosbuild3 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/8/execution/node/1107/log

daosbuild3 commented:
Test stage Functional on Leap 15.6 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/8/execution/node/1121/log

phender committed:
Merge branch 'master' into hendersp/DAOS-18019

phender committed:
Fix ubuntu.

Skip-func-test-leap15: false
Skip-func-test-el9: false
Test-tag: pr daily_regression

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

phender committed:
Use . instead of source

Skip-func-test-leap15: false
Skip-func-test-el9: false
Test-tag: pr daily_regression

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

phender committed:
Merge branch 'master' into hendersp/DAOS-18019

Skip-func-test-leap15: false
Skip-func-test-el9: false
Test-tag: pr daily_regression

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

phender committed:
Merge branch 'hendersp/DAOS-18019' of github.com:daos-stack/daos into hendersp/DAOS-18019

Skip-func-test-leap15: false
Skip-func-test-el9: false
Test-tag: pr daily_regression

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

daosbuild3 commented:
Test stage Functional on EL 9 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/12/execution/node/1109/log

daosbuild3 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/12/execution/node/1152/log

phender committed:
Merge branch 'master' into hendersp/DAOS-18019

phender committed:
Use sys.executable for python command.

Skip-func-test-leap15: false
Skip-func-test-el9: false
Test-tag: pr daily_regression

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

daosbuild3 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/14/execution/node/914/log

phender committed:
Merge branch 'master' into hendersp/DAOS-18019

phender committed:
Setup python virtual env on all ftest nodes.

Skip-func-test-leap15: false
Skip-func-test-el9: false
Test-tag: pr daily_regression

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

phender committed:
Merge branch 'master' into hendersp/DAOS-18019

phender committed:
Fix lint

Skip-func-test-leap15: false
Skip-func-test-el9: false
Test-tag: pr daily_regression

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

daosbuild3 commented:
Test stage Functional on Leap 15.6 completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-16919/17/testReport/

daosbuild3 commented:
Test stage Functional on EL 9 completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-16919/17/testReport/

daosbuild3 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/17/execution/node/1146/log

phender committed:
Install mpi4py via pip instead of RPM.

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

daosbuild3 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/19/execution/node/453/log

phender committed:
Revert RPM changes.

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

phender committed:
Merge branch 'master' into hendersp/DAOS-18019

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

phender committed:
Merge branch 'master' into hendersp/DAOS-18019

phender committed:
Fix dfuse/daos_build.py

Skip-func-test-leap15: false
Skip-func-test-el9: false
Test-tag: pr daily_regression

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

daosbuild3 commented:
Test stage NLT on EL 8.8 completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-16919/23/testReport/

daosbuild3 commented:
Test stage Functional on EL 9 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/23/execution/node/1131/log

daosbuild3 commented:
Test stage NLT on EL 8.8 completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-16919/24/testReport/

daosbuild3 commented:
Test stage Functional on Leap 15.6 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/26/execution/node/649/log

phender committed:
Merge branch 'master' into hendersp/DAOS-18019

Skip-func-test-leap15: false
Skip-func-test-el9: false
Test-tag: pr daily_regression

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

phender committed:
Merge branch 'master' into hendersp/DAOS-18019

Skip-func-test-leap15: false
Skip-func-test-el8: false
Test-tag: pr DfuseMUPerms test_dfuse_daos_build_wb

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

phender committed:
Fix copyrights

Skip-func-test-leap15: false
Skip-func-test-el8: false
Test-tag: pr DfuseMUPerms test_dfuse_daos_build_wb

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

daltonbohning commented on .github/workflows/ci2.yml:
```diff
@@ -27,11 +27,13 @@ jobs:
           - distro: ubuntu
             base: ubuntu
             with: ubuntu:noble
+            python: 3
```
Missed this one?
```suggestion
            python: 3.11
```

phender replied:
For ubuntu it's 3.  I previously tried using 3.11 but had to fix it in:
- https://github.com/daos-stack/daos/pull/16919/changes/9708ba62809a7fb5801a5271587e153a6d5f86da 
- https://github.com/daos-stack/daos/pull/16919/changes/dd2474552531b958070e823d20051192635873a5

daltonbohning replied:
I had to look this up. TL;DR is that Ubuntu is silly and does not provide specific python versions in their repos. We could enable a different repo but I don't know the security implications of that and we aren't really using Ubuntu enough right now to bother looking deeper.
`python3` on Ubuntu will give a different version of python depending on the OS, so I think we can just live with that for now.

daltonbohning commented on .github/workflows/landing-builds.yml:
```diff
@@ -49,19 +49,24 @@ jobs:
           - distro: ubuntu
             base: ubuntu
             with: ubuntu:noble
+            python: 3
```
```suggestion
            python: 3.11
```

phender replied:
See previous

daltonbohning commented on .github/workflows/landing-builds.yml:
```diff
@@ -105,11 +112,13 @@ jobs:
           - distro: ubuntu
             base: ubuntu
             with: ubuntu:noble
+            python: 3
```
```suggestion
            python: 3.11
```

phender replied:
For ubuntu I think this is appropriate

daltonbohning commented on .github/workflows/landing-builds.yml:
```diff
@@ -165,20 +175,25 @@ jobs:
           - distro: ubuntu
             base: ubuntu
             with: ubuntu:noble
+            python: 3
```
```suggestion
            python: 3.11
```

phender replied:
For ubuntu I think this is appropriate

daltonbohning commented on .github/workflows/landing-builds.yml:
```diff
@@ -241,18 +260,23 @@ jobs:
           - distro: ubuntu
             base: ubuntu
             with: ubuntu:24.04
+            python: 3
```
```suggestion
            python: 3.11
```

daltonbohning commented on ci/provisioning/post_provision_config_nodes_EL.sh:
```diff
@@ -28,5 +28,8 @@ group_repo_post() {
+    # Use a more recent python version for unit testing, this allows us to also test installing
+    # pydaos into virtual environments.
+    : "${PYTHON_VERSION:=}"
+    dnf -y install "python${PYTHON_VERSION}" "python${PYTHON_VERSION}-devel"
```
Defaulting this to an empty version means we might try to install `python` and `python-devel` which would give a different version depending on the os. Maybe just default to `3.11` here?
```suggestion
    : "${PYTHON_VERSION:=3.11}"
    dnf -y install "python${PYTHON_VERSION}" "python${PYTHON_VERSION}-devel"
```

phender replied:
Done

daltonbohning commented on ci/provisioning/post_provision_config_nodes_LEAP.sh:
```diff
@@ -23,4 +23,9 @@ distro_custom() {
+    # Use a more recent python version for unit testing, this allows us to also test installing
+    # pydaos into virtual environments.
+    : "${PYTHON_VERSION:=}"
+    dnf -y install "python${PYTHON_VERSION//./}" "python${PYTHON_VERSION//./}-devel"
```
```suggestion
    : "${PYTHON_VERSION:=3.11}"
    dnf -y install "python${PYTHON_VERSION//./}" "python${PYTHON_VERSION//./}-devel"
```

phender replied:
Done

daltonbohning commented on utils/rpms/packaging/Dockerfile.ubuntu:
```diff
@@ -21,11 +21,13 @@ RUN chmod +x /tmp/repo-helper.sh &&                 \
     rm -f /tmp/repo-helper.sh
 
 # Install basic tools
+ARG PYTHON_VERSION=3
```
How does `3` give `3.11`?

phender replied:
It doesn't :) The default python version is apparently based upon the ubuntu release, e.g.
- python 3.14 on 26.04
- python 3.12 on 24.04
- python 3.10 on 22.04
If we want a specific version for Ubuntu it looks like we'd need to add a repository that provides it - like Deadsnakes PPA - but there's no guaranteed security update timeline on a community-maintained repo.  Leaving the default PYTHON_VERSION=3 will just give us the default python version for the requested ubuntu release, which is probably OK for our current use cases.

daltonbohning commented on utils/scripts/install-ubuntu.sh:
```diff
@@ -62,8 +64,8 @@ apt-get install ${apt_get_install_args} \
     ruby \
-    python3-dev \
-    python3-venv \
+    python${PYTHON_VERSION}-dev \
```
Shouldn't this be like the others?
```suggestion
    python${PYTHON_VERSION//./}-dev \
```

phender replied:
That was only needed for Leap, so I think it's fine to take the literal variable string - like we do for EL*.

daltonbohning commented on ftest.sh:
```diff
@@ -132,6 +136,7 @@ if ! ssh -A $SSH_KEY_ARGS ${REMOTE_ACCT:-jenkins}@"${nodes[0]}" \
      STAGE_NAME=\"$STAGE_NAME\"
      DAOS_HTTPS_PROXY=\"$_DAOS_HTTPS_PROXY\"
      DAOS_NO_PROXY=\"$_DAOS_NO_PROXY\"
+     DAOS_FTEST_VENV=\"$DAOS_FTEST_VENV\"
```
Don't need to also pass PYTHON_VERSION here like above in this file?

phender replied:
We shouldn't need it anymore. I pushed an update earlier today that cleans up the `ftest/scripts/setup_nodes.sh` and `ftest/scripts/main.sh` usage:
- `setup_nodes.sh` (run on all hosts) will create the `$DAOS_FTEST_VENV` virtual environment using `$PYTHON_VERSION`
- `main.sh` (run on the launch node) will source the `$DAOS_FTEST_VENV` virtual environment and run launch.py

phender committed:
Cleanup ftest venv setup

Skip-func-test-leap15: false
Skip-func-test-el8: false
Test-tag: pr DfuseMUPerms test_dfuse_daos_build_wb

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

phender committed:
Applying feedback

Skip-func-test-leap15: false
Skip-func-test-el8: false
Test-tag: pr DfuseMUPerms test_dfuse_daos_build_wb

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>

daosbuild3 commented:
Test stage Functional on EL 9 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/33/execution/node/1194/log

phender commented:
> Test stage Functional on EL 9 completed with status FAILURE. https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/33/execution/node/1194/log

This failed because https://github.com/daos-stack/pipeline-lib/pull/515 has not been landed yet.

daosbuild3 commented:
Test stage Functional on SLES 15 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/34/execution/node/1193/log

daosbuild3 commented:
Test stage Functional on EL 9 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/34/execution/node/1213/log

daosbuild3 commented:
Test stage Functional on EL 9 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/35/execution/node/578/log

daosbuild3 commented:
Test stage Functional on SLES 15 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/36/execution/node/1146/log

daosbuild3 commented:
Test stage Functional on SLES 15 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/37/execution/node/1156/log

daosbuild3 commented:
Test stage Functional on Leap 15 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/38/execution/node/594/log

daosbuild3 commented:
Test stage Functional on SLES 15 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/38/execution/node/663/log

daosbuild3 commented:
Test stage Functional Hardware Medium MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/38/execution/node/913/log

phender commented:
> Test stage Functional Hardware Medium MD on SSD completed with status FAILURE. https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/38/execution/node/913/log

Known issue https://daosio.atlassian.net/browse/DAOS-18218

phender commented:
> Test stage Functional on SLES 15 completed with status FAILURE. https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16919/38/execution/node/663/log

Lost connection to brd-102vm08 during testing, e.g.
```
ssh: connect to host brd-102vm08 port 22: No route to host
````

phender committed:
Merge branch 'master' into hendersp/DAOS-18019

Fix merge conflict with copyright line in daos_server_helper.py

Skip-unit-tests: true
Skip-test: true

Signed-off-by: Phil Henderson <phillip.henderson@hpe.com>
